### PR TITLE
Touch Bug #1416256: React checkbox table widget

### PR DESCRIPTION
**Pull request by phlax**
Adds a react checkbox table, where you can de/select all and it can submit the visibly selected

### CODE REVIEW

File: pontoon/static/js/widgets/table.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        const all = (
+            checked.size > 0
+                && checked.size === this.visible.filter(v => v !== '').length);
+        return {all, checked};
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            this.visible.length = 0;
+            this.setState({checked: new Set()});
```
Comment (julen) [inline pontoon/static/js/widgets/table.js]: From the [docs on `componentWillUpdate`](https://reactjs.org/docs/react-component.html#componentwillupdate):
> If you need to update `state` in response to props changes, use `componentWillReceiveProps()` instead

File: pontoon/static/js/widgets/table.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        }
+    }
+
+    handleCheckboxClick (evt) {
+        // adds/removes paths for submission
+        let {checked} = this.state;
+        checked = new Set(checked);
+        evt.target.checked ? checked.add(evt.target.name) : checked.delete(evt.target.name);
+        this.setState({checked});
```
Comment (julen) [inline pontoon/static/js/widgets/table.js]: Note how setting the new state depending on the current value of `state` should be done [via the function argument](https://reactjs.org/docs/faq-state.html#how-do-i-update-state-with-values-that-depend-on-the-current-state).

File: pontoon/static/js/widgets/table.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        return this.setState({checked});
+    }
+
+    async handleSubmit (evt) {
+        // after emitting handleSubmit to parent with list of currently
+        // checked, clears the checkboxes
+        evt.preventDefault();
+        const {checked} = this.state;
+        await this.props.handleSubmit({data: [...checked]});
+        await this.setState({checked: new Set()});
```
Comment (julen) [inline pontoon/static/js/widgets/table.js]: `setState()` doesn't return a `Promise` (in fact, it doesn't return anything), and `await` is not needed. All `return` statements for `this.setState({...})` are unnecessary too.

File: pontoon/static/js/widgets/table.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    renderTable () {
+        return (
+            <ReactTable
+               defaultPageSize={this.defaultPageSize}
+               className="-striped -highlight"
+               data={this.props.data}
+               onPageChange={this.handleTableChange}
+               onPageSizeChange={this.handleTableResize}
+               onSortedChange={this.handleTableSortChange}
+               columns={this.columns} />);
```
Comment (julen) [inline pontoon/static/js/widgets/table.js]: If I'm not mistaken, `this.columns` is undefined.

File: pontoon/static/js/widgets/table.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        const checked = this.pruned
+        const all = (
+            checked.size > 0
+                && checked.size === this.visible.filter(v => v !== '').length);
+        return {all, checked};
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            this.visible.length = 0;
```
Comment (julen) [inline pontoon/static/js/widgets/table.js]: This one is surprising: it changes the length of an array. At the same time, `visible` gets no value besides the initial one (`[]`).

Comment (phlax) [reply pontoon/static/js/widgets/table.js]: the design of this widget is to do exactly that - ie set the length of the array according to the number of visible rows there are. The slots are filled as the rows are rendered.

File: pontoon/static/js/widgets/table.js (hunk 4 above)
Comment (phlax) [reply pontoon/static/js/widgets/table.js]: hmm, yep, this was due to this component still being inherited from - i was on the way to changing that when splitting out this PR. @julen thanks for this review, its been really helpful

File: tests/js/widgets/tables/checkbox.js [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+    let result = table.instance().selected;
+
+    // default is for there to be no checked
+    expect(result).toEqual({all: false, checked: new Set([])});
+
+    // set some visible items
+    table.instance().visible = [4, 5, 6];
+
+    // set checked items
+    await table.instance().setState({checked: new Set([1, 2, 3])});
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: This `await` does nothing.

File: tests/js/widgets/tables/checkbox.js [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    // tests the CheckboxTable componentWillReceiveProps method
+
+    const data = [23];
+    const table = shallow(<CheckboxTable data={data} />);
+
+    // visible is the (tracked) currently visible resources
+    table.instance().visible = [1, 2, 3];
+
+    // lets setState with some "checked" data
+    await table.setState({checked: new Set(['a', 'b', 'c'])});
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: This `await` does nothing.

File: tests/js/widgets/tables/checkbox.js [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+
+    // calling with checked=false removes the item
+    evt.target.checked = false;
+    table.instance().handleCheckboxClick(evt);
+    checked.delete(23);
+    expect(table.state()).toEqual({checked});
+});
+
+
+test('CheckboxTable componentWillReceiveProps', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

File: tests/js/widgets/tables/checkbox.js [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+    // from the set, and evt.preventDefault was fired again
+    expect(handleSubmit.mock.calls[1]).toEqual([{"data": [1, 2, 3]}]);
+    expect(evt.preventDefault.mock.calls[1]).toEqual([]);
+
+    // state.checked got cleared
+    expect(table.state().checked).toEqual(new Set());
+});
+
+
+test('CheckboxTable handleCheckboxClick', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

File: tests/js/widgets/tables/checkbox.js [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    expect(table.instance().visible).toEqual(
+        [23, 43, undefined, undefined, undefined]);
+
+    // state.checked gets the common items
+    let stateFunc = table.instance().setState.mock.calls[0][0];
+    expect(stateFunc(table.state())).toEqual({"checked": new Set([23, 43])});
+});
+
+
+test('CheckboxTable prune', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

File: tests/js/widgets/tables/checkbox.js [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+
+    table.instance().handleTableSortChange();
+
+    // setState is called with a checked set containing just 23
+    let stateFunc = table.instance().setState.mock.calls[0][0]
+    expect(stateFunc(table.state())).toEqual({"checked": new Set([23])});
+});
+
+
+test('CheckboxTable handleTableChange', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

File: tests/js/widgets/tables/checkbox.js [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+    // calling handleSelectAll will clear any checked items
+    // if they are not visible
+    table.instance().state.checked = new Set([4, 5]);
+    table.instance().handleSelectAll()
+    stateFunc = table.instance().setState.mock.calls[2][0]
+    expect(stateFunc(table.state())).toEqual({"checked": new Set()});
+});
+
+
+test('CheckboxTable handleTableSortChange', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

File: tests/js/widgets/tables/checkbox.js [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+    // if all are selected then indeterminate is false, and checked is true
+    selected.all = true;
+    checkbox = table.instance().renderSelectAllCheckbox();
+    expect(checkbox.props.checked).toBe(true);
+    expect(checkbox.props.indeterminate).toBe(false);
+    expect(checkbox.props.onClick).toBe(table.instance().handleSelectAll);
+});
+
+
+test('CheckboxTable handleSelectAll', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

File: tests/js/widgets/tables/checkbox.js [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+        {original: ['FOO'], pageSize: 5});
+
+    // the checkbox's props are set appropriately
+    expect(checkbox.props.checked).toBe(true);
+    expect(checkbox.props.name).toBe('FOO');
+    expect(checkbox.props.onChange).toBe(table.instance().handleCheckboxClick);
+});
+
+
+test('CheckboxTable selected', async () => {
```
Comment (adngdb) [inline tests/js/widgets/tables/checkbox.js]: No need for this function to be `async`.

**Pull request merged**